src/cascadia/TerminalCore/TerminalSelection.cpp

```diff
@@ -235,92 +73,173 @@ const bool Terminal::IsCopyOnSelectActive() const noexcept
 // Method Description:
-// - Select the entire row of the position clicked
+// - Record the position of the beginning of a selection
 // Arguments:
 // - position: the (x,y) coordinate on the visible viewport
-void Terminal::TripleClickSelection(const COORD position)
+void Terminal::SetSelectionAnchor(const COORD viewportPos)
 {
-    SetSelectionAnchor({ 0, position.Y });
-    SetEndSelectionPosition({ _buffer->GetSize().RightInclusive(), position.Y });
+    _selectionPivot = _ConvertToBufferCell(viewportPos);
+    _buffer->GetSize().Clamp(_selectionPivot);
+
+    _multiClickSelectionMode = SelectionExpansionMode::Cell;
 
-    _multiClickSelectionMode = SelectionExpansionMode::Line;
+    // Unlike MultiClickSelection(), we need to set
+    // these vars BEFORE SetSelectionEnd()
+    _allowSingleCharSelection = (_copyOnSelect) ? false : true;
+    _selectionStart = _selectionPivot;
+
+    SetSelectionEnd(viewportPos);
 }
 
 // Method Description:
-// - Record the position of the beginning of a selection
+// - Update selection anchors when dragging to a position
+// - based on the selection expansion mode
 // Arguments:
-// - position: the (x,y) coordinate on the visible viewport
-void Terminal::SetSelectionAnchor(const COORD position)
+// - viewportPos: the (x,y) coordinate on the visible viewport
+// - expansionModeOverwrite: overwrites the _multiClickSelectionMode for this function call. Used for ShiftClick
+void Terminal::SetSelectionEnd(const COORD viewportPos, std::optional<SelectionExpansionMode> expansionModeOverwrite)
 {
-    _selectionAnchor = position;
+    auto textBufferPos = _ConvertToBufferCell(viewportPos);
+    _buffer->GetSize().Clamp(textBufferPos);
 
-    // include _scrollOffset here to ensure this maps to the right spot of the original viewport
-    THROW_IF_FAILED(ShortSub(_selectionAnchor.Y, gsl::narrow<SHORT>(_scrollOffset), &_selectionAnchor.Y));
+    // if this is a shiftClick action, we need to overwrite the _multiClickSelectionMode value (even if it's the same)
+    // Otherwise, we may accidentally expand during other selection-based actions
+    const auto shiftClick = expansionModeOverwrite.has_value();
+    _multiClickSelectionMode = expansionModeOverwrite.has_value() ? *expansionModeOverwrite : _multiClickSelectionMode;
 
-    // copy value of ViewStartIndex to support scrolling
-    // and update on new buffer output (used in _GetSelectionRects())
-    _selectionVerticalOffset = gsl::narrow<SHORT>(ViewStartIndex());
+    switch (_multiClickSelectionMode)
+    {
+    case SelectionExpansionMode::Line:
+        _ChunkSelectionByLine(textBufferPos, shiftClick);
+        break;
+    case SelectionExpansionMode::Word:
+        _ChunkSelectionByWord(textBufferPos, shiftClick);
+        break;
+    case SelectionExpansionMode::Cell:
+    default:
+        _ChunkSelectionByCell(textBufferPos, shiftClick);
+
+        if (_copyOnSelect && !_IsSingleCellSelection())
+        {
+            _allowSingleCharSelection = true;

```

Comment:
comment note: "// moving the endpoint of what used to be a single cell selection allows the user to drag back and select just one cell"